backend/main.py

```
@@ -38,30 +74,183 @@
+UPLOAD_LIMIT_PER_USER = 5  # max uploads per user per hour
+UPLOAD_LIMIT_PER_IP = 10  # max uploads per IP per hour
+
+# Image processing cache to avoid duplicate API calls
+# Replaced custom cache with async_lru for better performance and memory management
+
+def check_upload_limits(identifier: str, limit: int) -> None:
+    """
+    Check if the user/IP has exceeded upload limits using thread-safe cache.
+    """
+    current_uploads = user_upload_cache.get(identifier) or []
+    now = datetime.now()
```

Ruff (0.14.14)

[error] Undefined name `datetime`

(F821)